daemon/gamemode-context.c

```diff
@@ -178,6 +193,132 @@ void game_mode_context_destroy(GameModeContext *self)
+		gov_str = gov_config_str[0] != '\0' ? gov_config_str : self->initial_cpu_mode;
+		break;
+
+	case GAME_MODE_GOVERNOR_DESIRED:
+		config_get_desired_governor(self->config, gov_config_str);
+		gov_str = gov_config_str[0] != '\0' ? gov_config_str : "performance";
+		break;
+
+	case GAME_MODE_GOVERNOR_IGPU_DESIRED:
+		config_get_igpu_desired_governor(self->config, gov_config_str);
+		gov_str = gov_config_str[0] != '\0' ? gov_config_str : "powersave";
+		break;
+
+	default:
+		assert(!"Invalid governor requested");
+	}
+
+	const char *const exec_args[] = {
+		"/usr/bin/pkexec", LIBEXECDIR "/cpugovctl", "set", gov_str, NULL,
+	};
+
+	LOG_MSG("Requesting update of governor policy to %s\n", gov_str);
+	int ret = run_external_process(exec_args, NULL, -1);
+	if (ret != 0) {
+		LOG_ERROR("Failed to update cpu governor policy\n");
+		return ret;
+	}
+
+	/* Update the current govenor only if we succeed at setting govenors. */
+	self->current_govenor = gov;
+
+	return 0;
+}
+
+static void game_mode_check_igpu_energy(GameModeContext *self)
+{
+	float threshold = config_get_igpu_power_threshold(self->config);
+	if (threshold > 10000) {
+		/* There's no way the GPU is using 10000x the power.  This lets us
+		 * short-circuit if the config file specifies an invalid threshold
+		 * and we want to disable the iGPU heuristic.
+		 */
+		return;
+	}
+
+	pthread_rwlock_wrlock(&self->rwlock);
+
+	/* We only care if we're not in the default governor */
+	if (self->current_govenor == GAME_MODE_GOVERNOR_DEFAULT)
+		goto unlock;
+
+	if (!self->igpu_energy_valid) {
+		/* If we failed to get valid energies when we first entered game
+		 * mode, we're not likely to get valid values now.  Integrated
+		 * GPUs aren't a thing you hotplug.
+		 */
+		goto unlock;
+	}
+
+	uint32_t cpu_energy_uj, igpu_energy_uj;
+	if (!get_cpu_energy_uj(&cpu_energy_uj) ||
+	    !get_igpu_energy_uj(&igpu_energy_uj)) {
+		/* We've already succeeded at getting power information once so
+		 * failing here is possible but very unexpected. */
+		self->igpu_energy_valid = false;
+		LOG_ERROR("Failed to get CPU and iGPU power data\n");
+		goto unlock;
+	}
+
+	/* Overflow is possible here.  However, that would simply mean that
+	 * the HW counter has overflowed and us wrapping around is probably
+	 * the right thing to do.  Wrapping at 32 bits is exactly what the
+	 * Linux kernel's turbostat utility does so it's probably right.
+	 */
+	uint32_t cpu_energy_delta_uj = cpu_energy_uj - self->last_cpu_energy_uj;
+	uint32_t igpu_energy_delta_uj = igpu_energy_uj - self->last_igpu_energy_uj;
+	self->last_cpu_energy_uj = cpu_energy_uj;
+	self->last_igpu_energy_uj = igpu_energy_uj;
+
+	if (cpu_energy_delta_uj == 0) {
+		LOG_ERROR("CPU reported no energy used\n");
+		goto unlock;
+	}
+	if (igpu_energy_delta_uj == 0) {
+		LOG_ERROR("Integrated GPU reported no energy used\n");
+		goto unlock;
+	}
+
+	double ratio = (double)igpu_energy_delta_uj /
+	               (double)cpu_energy_delta_uj;
+	if (ratio > threshold) {
```

Comment:
It's possible but I think ultimately unlikely.  Once you go above the threshold, it switches to the powersave governor at which point the power used by the CPU is likely to significantly decrease making the ratio even larger so there's a natural bias in the system which does exactly what you want.  Unfortunately, I'm not sure how reliable that is.

Reply:
Ah, that makes sense. It might be prudent to try and graph how long the response time on that switch is, but given our current polling rate is 5 seconds by default we're likely in the clear.